drivers/sm_pwm_01c/sm_pwm_01c.c

```diff
@@ -0,0 +1,211 @@
+#ifdef MODULE_SM_PWM_01C_MA
+    _circ_buf_push(&dev->values.tsp_circ_buf, tsp);
+    _circ_buf_push(&dev->values.tlp_circ_buf, tlp);
+#else
+    dev->values.data.mc_pm_10 =
+        (uint16_t)((tlp + (uint32_t)(CONFIG_SM_PWM_01C_EXP_WEIGHT - 1) *
+                    dev->values.data.mc_pm_10) / CONFIG_SM_PWM_01C_EXP_WEIGHT);
+    dev->values.data.mc_pm_2p5 =
+        (uint16_t)((tsp + (uint32_t)(CONFIG_SM_PWM_01C_EXP_WEIGHT - 1) *
+                    dev->values.data.mc_pm_2p5) / CONFIG_SM_PWM_01C_EXP_WEIGHT);
+#endif
+
+    /* reset lpo */
+    dev->values.tlp_lpo = 0;
+    dev->values.tsp_lpo = 0;
+}
+
+static void _tsp_pin_cb(void *arg)
+{
+    sm_pwm_01c_t *dev = (sm_pwm_01c_t *)arg;
+    uint32_t now = ztimer_now(ZTIMER_USEC);
+
+    if (gpio_read(dev->params.tsp_pin) == 0) {
+        dev->values.tsp_start_time = now;
+    }
+    else {
+        dev->values.tsp_lpo += (now - dev->values.tsp_start_time);
+    }
+}
+
+static void _tlp_pin_cb(void *arg)
+{
+    sm_pwm_01c_t *dev = (sm_pwm_01c_t *)arg;
+    uint32_t now = ztimer_now(ZTIMER_USEC);
+
+    if (gpio_read(dev->params.tlp_pin) == 0) {
+        dev->values.tlp_start_time = now;
+    }
+    else {
+        dev->values.tlp_lpo += (now - dev->values.tlp_start_time);
+    }
+}
+
+int sm_pwm_01c_init(sm_pwm_01c_t *dev, const sm_pwm_01c_params_t *params)
+{
+    dev->params = *params;
+
+    /* set up irq */
+    if (gpio_init_int(dev->params.tsp_pin, GPIO_IN_PU, GPIO_BOTH, _tsp_pin_cb,
+                      dev) < 0) {
+        DEBUG("[sm_pwm_01c] init_int of tsp_pin failed [ERROR]\n");
+        return -EIO;
+    }
+    if (gpio_init_int(dev->params.tlp_pin, GPIO_IN_PU, GPIO_BOTH, _tlp_pin_cb,
+                      dev) < 0) {
+        DEBUG("[sm_pwm_01c] init_int of tlp_pin failed [ERROR]\n");
+        return -EIO;
+    }
+
+    /* setup timer */
+    _timer.callback = _sample_timer_cb;
+    _timer.arg = dev;
+
+#ifdef MODULE_SM_PWM_01C_MA
+    memset(&dev->values.tsp_circ_buf, 0, sizeof(circ_buf_t));
+    memset(&dev->values.tlp_circ_buf, 0, sizeof(circ_buf_t));
+#endif
+
+    return 0;
+}
+
+void sm_pwm_01c_start(sm_pwm_01c_t *dev)
+{
+    assert(dev);
+    /* reset old values */
+    memset((void *)&dev->values, 0, sizeof(sm_pwm_01c_values_t));
+    /* enable irq and set timer */
+    ztimer_set(ZTIMER_USEC, &_timer, CONFIG_SM_PWM_01C_SAMPLE_TIME);
+    gpio_irq_enable(dev->params.tsp_pin);
+    gpio_irq_enable(dev->params.tlp_pin);
+    DEBUG("[sm_pwm_01c] started average measurements\n");
+}
+
+void sm_pwm_01c_stop(sm_pwm_01c_t *dev)
+{
+    assert(dev);
+    /* disable irq and remove timer */
+    ztimer_remove(ZTIMER_USEC, &_timer);
+    gpio_irq_disable(dev->params.tsp_pin);
+    gpio_irq_disable(dev->params.tlp_pin);
+    DEBUG("[sm_pwm_01c] stopped average measurements\n");
+}
+
+void sm_pwm_01c_read_data(sm_pwm_01c_t *dev, sm_pwm_01c_data_t *data)
+{
+    assert(dev);
+#ifdef MODULE_SM_PWM_01C_MA
+    unsigned int state = irq_disable();
```

Comment:
The `irq_disable`/`irq_enable` could be placed just once outside the `ifdef` right?